### PR TITLE
Add the notification rules.

### PR DESCRIPTION
**Pull request by Elv13**
This is the final "main" PR in the notification rewrite. It replaces the old "preset" concept with a full `awful.rules` like system to control the notification. It can be used to set the color, ignore them, set custom widget templates, etc. The basic usage is:

```lua
   ruled.notification.connect_signal("request::rules", function()
       -- Add a red background for urgent notifications.
       ruled.notification.append_rule {
           rule       = { urgency = "critical" },
           properties = { bg = "#ff0000", fg = "#ffffff", timeout = 0 }
       }

       -- Or green background for normal ones.
       ruled.notification.append_rule {
           rule       = { urgency = "normal" },
           properties = { bg      = "#00ff00", fg = "#000000"}
       }
   end)
```

Please note that this is currently in a new module called "ruled". I prefer this over adding `naughty.rules` since I also plan to expand the rules to tags, screens and the clipboard. `awful.rules` itself would be moved to this new module. If someone has a better name, please come forward.

The content of this PR is less or more final. ~~The tests are lacking and a bit broken~~. The NEWS.md also needs to updated to reflect the entire rewrite effort. I don't plan to merge this in the immediate future. It will co-exist for a while with some other PRs to introduce the other components of the new `ruled` module.

### CODE REVIEW

**Comment by Elv13** [comment]
I added more tests. As for the "build failure", it is because the build directory isn't clean when iterating each commit. Because `ruled/notification.lua` is modified once, the "old" copy is left in the build tree after the builder finish building the last commit and start iterating each commit from the first. One way to do this is to add another commit to use the `@rulelib` and other new tags in the last commit. However that's a hack a "someone" will complain how ugly that it, so I leave it as-is for now.

@actionless Can you try this new API and see if it fits your needs? You wondered how easy some tasks were going to be. Time to try it!

**Comment by SethBarberee** [comment]
@Elv13 I can give this a run for a bit on my machine and give feedback. Any specific test cases you want tested? I can give some feedback within the week

**Comment by Elv13** [comment]
Sure, go hard at it. There is no specific things to try. This is probably
the most useful for filtering and to replace the old preset API. And of
course for custom template.

**Comment by SethBarberee** [comment]
```
    ruled.notification.append_rule {
        rule       = { app_name = 'Spotify' },
        properties = { 
            bg      = '#000000', 
            fg = '#00ff00',
            append_actions = {
               naughty.action {
                   name = 'Snooze (5m)',
               },
               naughty.action {
                   name = 'Snooze (15m)',
               },
               naughty.action {
                   name = 'Snooze (1h)',
               },
            },
        }
    }
```
Good news: 
Was able to add actions to Spotify... you are a genius, Elv. Happy to see that my reddit post in regards to notifications special to spotify is possible now :smile: 

Bad news (only in `naughty.layout.box` formatted notifications):
While the said notification is green, actions are default foreground.
**This doesn't occur in the old style as actions are changed to said green**

**Comment by Elv13** [comment]
> While the said notification is green, actions are default foreground.

Yeah, `beautiful` has precedence unless you set a specific `fg` for the actions. This might not have been the best design decision. I will see what I can do about it. You *can* make rules to color the actions, but they should probably follow the whole notification fg unless explicitly overridden rather than use their own fallback when a notification `fg` is set, but not an action `fg`.

**Comment by actionless** [comment]
could you please rebase it so i could merge it to my local branch to test

**Comment by Elv13** [comment]
> could you please rebase it so i could merge it to my local branch to test

Done. I pressed the "resolve conflicts" button in the GUI. It seems to have worked.

**Comment by SethBarberee** [comment]
With #2895 merged, is this still WIP? Wanted to make sure before I deep dive and do more testing/review of it. Looking forward to this PR.

**Comment by Elv13** [comment]
>  is this still WIP?

Yes, there is a bunch of problems:

 * The multi-screen support for the `box` notification isn't behaving properly
 * Some signal signature are incompatible with my API standardization work I just merged (facepalm)
 * The icon update for Spottily notification is still broken.
 * It doesn't have a permission per-client block the notification using the new permission framework

Beside, it is working and not much will change beside the signal format.

**Comment by Elv13** [comment]
I consider this PR to no longer be WIP/RFC. While I didn't fix the spotify bug because I don't have a setup to reproduce it, I think otherwise, it is generally useful and stable enough to merge and enable by default. At this point, most "major" configs as seen on `/r/unixporn` and our screenshot page migrated to the new notification API. I guess it works for them, so lets make this a bit easier for newcomers to use. I started this notification project 3 years ago (almost, technically in June). It has been a long journey to get this merged. Hundreds of commits are related to the groundwork leading to this PR getting merged...

File: lib/ruled/notification.lua [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+--- Append some actions to a notification.
+--
+-- Using `actions` directly is destructive since it will override existing
+-- actions.
+--
+-- @clientruleproperty append_actions
+-- @param table
+
+--- Set a fallback timeout the notification has no explicit timeout.
```
Comment (SethBarberee) [inline lib/ruled/notification.lua]:
```suggestion
--- Set a fallback timeout when the notification doesn't have an explicit timeout.
```

**Comment by Aire-One** [comment]
I never have desktop notification so I never took the time to implement your new API in my setup but I silently followed each step of the development. It was a long journey.

This is great to see it makes its way to default `rc.lua`. ea79035 is the final step of this long running project. 

Congrats! 🎊

**Pull request merged**